### PR TITLE
Implement diff deduplication

### PR DESCRIPTION
**Pull request by pierd**
- filters out `WorldChange::Set` operations if they already existed in the previous diff
- skips sending `WorldChange::Set` operations for removed entities/components (previously we would send them but they would silently fail since they are applied as last ones)
- changes `WorldChange::Set` into `WorldChange::SetComponents` for updating multiple components of the same entity at once
- removes `create_revert` flag from `WorldDiff` and `WorldChange` - dead code and I didn't want to create an extra implementation for `WorldChange::SetComponents`

With these changes `afps` uses ~100kB/s instead of ~4000kB/s when all blocks are destroyed.

### CODE REVIEW

File: crates/network/src/diff_serialization.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                    // we only populate components_to_drop for SetComponents changes
+                    unreachable!();
+                }
+            }
+            keep
+        });
+        self.last_diff = new_diff;
+        diff
+    }
+}
```
Comment (pierd) [inline crates/network/src/diff_serialization.rs:57]: New module. I plan to put more serialisation specific code here (for #627 and #628).

**Review by FredrikNoren — APPROVED** [review]
Neat! The only thing I want to request is an update to the book with info about this so people know what to expect

File: docs/src/reference/networking.md [hunk 2]
```diff
@@ -16,6 +16,12 @@ The client is fundamentally designed around runtime flexibility of logic, which
 
 If on 0.2 or above, consult the [clientside](https://github.com/AmbientRun/Ambient/blob/main/guest/rust/examples/intermediate/clientside/ambient.toml) example to see how to define networked components.
 
+### Entities synchronization
+
+The Ambient runtime synchronizes entities using a diff-based approach. The server sends a `WorldDiff` to the client, which contains a list of entities to spawn and despawn, and components to add, update, and remove. Note that some operations might be batched for performance or not included in the update sent to the clients if they effectively don't change any values. For example adding 0 to a number or changing a boolean to false and back to true within the same frame might not emit an update and might not trigger a `change_query`. We recommend using messaging if such events are important to your game.
+
+Currently the client applies the changes to its local world as soon as they are received.
+
```
Comment (pierd) [inline docs/src/reference/networking.md:24]: @FredrikNoren how does it sound?

File: crates/ecs/src/stream.rs [hunk 3]
```diff
@@ -204,38 +190,37 @@ pub struct FrozenWorldDiff {
...
-                        continue;
+        for change in diffs.iter().flat_map(|diff| diff.changes.iter()) {
+            if let WorldChange::SetComponents(entity_id, entity) = change {
+                if let Some(idx) = set_idx.get(entity_id) {
+                    if let Some(WorldChange::SetComponents(_, existing_entity)) =
+                        changes.get_mut(*idx).map(Cow::to_mut)
+                    {
+                        existing_entity.merge(entity.clone());
+                    } else {
+                        panic!("Expected SetComponents");
```
Comment (philpax) [inline crates/ecs/src/stream.rs:206]: Can you extend this panic/add a comment about the invariants that would prevent this panic from firing? (Or if it *can* fire, can you add more information about what to look out for?)

**Review by philpax — APPROVED** [review]

File: crates/ecs/src/stream.rs (hunk 3 above)
Comment (pierd) [reply crates/ecs/src/stream.rs:206]: It should actually be `unreachable!()`. I'll change it and add a comment.

Comment (pierd) [reply crates/ecs/src/stream.rs:206]: https://github.com/AmbientRun/Ambient/commit/e3318819325dd26ce1f317b39d4e72723b9ca10e

**Pull request merged**